### PR TITLE
Add tool for checking licenses

### PR DESCRIPTION
**Pull request by Cypher1**
A step forward in addressing https://github.com/google-research/raksha/issues/101

### CODE REVIEW

**Comment by Cypher1** [comment]
Note, I've added a `bin` folder and a `bin/hooks` folder, following advice from https://stackoverflow.com/questions/3462955/putting-git-hooks-into-a-repository

I'm unsure if this is appropriate for our project, but thought it worth sketching out the rough shape we might need to make this work for us.

Looking forward to feedback.

File: check_licenses.sh
Comment (Cypher1) [reply check_licenses.sh]: Thanks Mark! I've updated it as a pre-commit hook.
I also:

- Switched to using Pathspecs to make things more managable (and added comments on each group of specs).
- Added support for a 'source' subcommand which checks all 'source' files (not just changed files)
- Added a subcommand for checking 'all' files, regardless of their filename (to use as a reference point)
- Added a help subcommand to make using the script easier (the USAGE string that it echos should be easy to find even if running `./scripts/hooks/pre-commit help` isn't obvious).
- Added licenses to the scripts I've added (amusingly I didn't originally think to include them).
- Moved (non-script) file changes into a second PR which I can land once this is merged.

You're right that this can be bypassed, but the intention is not to enforce anything, but help guide towards best practice.
Looking forward to your review :)

Aside:
It's worth noting that your previous work in #282 (referenced in #285), protects against missing licenses that could impact internal code. Unfortunately I couldn't see a good way to de-duplicate code here, but perhaps that is something your could advise on?

File: scripts/hooks/pre-commit
Comment (Cypher1) [inline scripts/hooks/pre-commit:64]: This assumes that any license header will include a link to the actual license... this may not be a good assumption, but seemed to work for our code so far.

File: scripts/hooks/pre-commit
Comment (Cypher1) [reply scripts/hooks/pre-commit:77]: Ah! Thanks, I'll find the first check in then. Thanks

Comment (Cypher1) [reply scripts/hooks/pre-commit:77]: > Interesting; why check for this rather than the following?
> 
> ```
> Copyright 20[0-9[0-9] Google LLC
> ```


I thought that would be better at first, but we have code licensed to other authors (e.g. Bazel).

I think you're right and I should switch to checking each of a set of accepted licenses rather than trying to match them all at once.

File: check_licenses.sh
Comment (markww) [reply check_licenses.sh]: I also don't see a good way to dedup the code. The difficulty in sharing code is the big reason why I think it's probably OK to have a commit hook only rather than trying to maintain this code in the presubmit as well. Agreed that guidance, rather than enforcement, is the goal here.

**Review by markww — APPROVED** [review]
LGTM modulo comments already left, feel free to get this in once those are addressed.

File: scripts/hooks/pre-commit
Comment (Cypher1) [reply scripts/hooks/pre-commit]: Ah, the files that have different licenses are exclusively in special cased files already. I've switched to using the regex you suggested!

**Comment by Cypher1** [comment]
@markww As this doesn't change anything internal, I think this should be merged here (and I do not currently have the permissions to do so). Please merge this when you can :)

Thanks

**Pull request merged**
